### ISSUE TITLE
Feature Request: Kenwood 710 COM port mode

### ISSUE THREAD

**Issue by jesshaas** [ge0rg/aprsdroid#154]
I know about the Kenwood GPS port mode but the COM serial port on the back of the head of the D710 seems even better for our uses. When in APRS mode it outputs every received APRS packet. samples of data output on the com port when the TNC is in APRS mode are below:
cmd:cmd:K6TZ-11>APTW14,K6ERN,W6SCE-10*:_01220044c273s000g000t043r...p...P...h..b10087tU2k
DU1XO>S4PRVX,KG6JOA,WIDE1,W6SCE-10*:`-P3l#)>/>"8"}DU1XO/W6 Mabuhay!=
N7JY>APT311,W6SCE-10*,WIDE2-1:>n7jy@excite.com00g000t043r...p...P...h..b10083tU2k

Any chance of adding a setting to decode these? When I put aprsdroid into KISS mode it does not do it. aprsdroid works great with the radio in TNC mode but the APRS interface on the radio is great and does a number of things aprsdroid doesn't so I would like to keep it functional.

As far as transmitting packets goes I don't know if there is any way to do it with the radio in APRS mode but I actually have an additional bluetooth serial adapter hooked up to the PC port on the radio for rig control. It should be possible for aprsdroid to use this in order to switch the TNC into TNC mode, send a packet and then switch back to APRS mode in order to send messages, etc.... Implementing rig control could also be useful in the future for implementing things like APRS satellite support. Would be cool to have it automatically switch to satellite frequencies when they are overhead, adjust for doppler shift, etc... when out in the middle of nowhere without APRS infrastructure. 

Thanks,
   Jess
   KM6GVW

**Comment by jesshaas** [comment]
I may have been wrong about being able to change TNC mode from the PC port. You can change into TNC mode from APRS mode by sending TN 2,0 over the COM port but once in TNC mode I haven't been able to figure out how to exit out of it without pressing a button on the front panel. Receive should be easy, for sending messages the user might just have to press the button on the front of the radio to switch back into APRS mode. Not as nice but not a huge deal for most use.

**Comment by jesshaas** [comment]
I found a command that works referenced on http://www.trinityos.com/HAM/CentosDigitalModes/misc/D710-TNC-setup.txt

So when the TNC is in APRS mode the following command switches to packet mode:
TN 2,0

When in packet mode the following two commands are required to switch back to APRS mode:
TC 1
TN 1,0

References:
https://github.com/LA3QMA/TM-V71_TM-D710-Kenwood/blob/master/commands/TN.md
https://github.com/LA3QMA/TM-V71_TM-D710-Kenwood/blob/master/commands/TC.md

So should be able to switch seamlessly between receiving and transmitting while keeping the radios APRS interface working.

**Comment by dranch** [comment]
Yes, the D710 can be given commands to switch between APRS posit mode and it's internal TNC but it's not very quick about it. I don't know what Jess's desired application is here but switching from APRS mode into command mode, enable KISS, have APRSDroid initialize the KISS stack, send a UI packet, exit KISS mode, and return back to to APRS mode might take something like 3-5 seconds.  
Btw, Xastir for Linux supports decoding of these CMD mode APRS strings so examples can be found there.

**Comment by jesshaas** [comment]
I don't really see slow switching as an issue for most use cases. Personally I want to have a tablet hooked up to my 710 that I can use for APRS map view and to send messages while maintaining the 710s APRS interface. An occasional missed packet while switching just isn't really an issue.

**Comment by jesshaas** [comment]
I wrote up a little writeup on how to setup bluetooth on the COM port for cheap. http://jesshaas.com/projects/d710bt/ Also works for the GPS port which I am currently using but I really want to get this mode working through the COM port so I can get full functionality.

**Comment by ge0rg** [comment]
Hey @jesshaas, I've seen your post on the D7x0 ML as well. Very nice writeup, and sorry I haven't had time to comment on this topic earlier on.

To summarize the ideas, it looks like there are three things here:

1. Make APRSdroid decode the "cmd:"*X prefixed APRS packets in TNC2 format. 

   It is not hard to strip repeating "cmd:" from the beginning of packets to parse, however I remember there are other issues as well - essentially, in APRS mode the D7xx does not do proper logging, it is just a byproduct. The main problem arises when the D7xx emits a packet of its own, IIRC there is some other garbage on the serial port in that case.

2. Transmitting packets from APRSdroid in parallel to the Kenwood APRS module.

   This is a real can of worms: APRS is a stateful protocol (think of message ACKs and SmartBeaconing), and having both APRSdroid and the D7xx operate on the same SSID will cause issues for sure. Operating on a different SSID will cause other problems, like double the position traffic and/or possible issues with message routing to the right SSID.

3. Rig control

   This is a topic I haven't looked into at all yet. It would be great to have another developer jump in and provide an Android Intent interface, similar to https://aprsdroid.org/api/ to perform configuration tasks. It could be easily added into APRSdroid for single-tap-tuning to a frequency etc. I'd rather not have APRSdroid contain two different (Bluetooth?) links to the radio, one for the TNC and another for rig control, and I think the runtime-switch-to-KISS idea is rather hacky and error prone.

**Comment by jesshaas** [comment]
Even if only #1 was implemented it would still provide far more functionality than the GPS waypoint mode currently provides and allow for full radio connectivity using only two bluetooth modules(currently it needs 3). The cmd: part of the output wasnt actually there for every packet, most of the time just the packet was being output. i think i was connected when i beaconed and i dont remember garbage being output but i could be wrong. Even if there is garbage it doesnt seem difficult to ignore.

The only real need I see for aprsdroid transmitting in this mode is for sending messages. Typing messages on the radio is not nearly as nice as on the phone. That said I know there are lots of people that would really like to use their phones keyboard to send messages through their D710 without losing the radios built in APRS functionality.

Rig control would be cool for other things such as tuning to frequencies people are on and aprs satellite tracking but I was actually wrong about it being needed for this. You can switch back and forth between APRS and TNC modes via the TNC COM port. So it can all be done via a single bluetooth connection.

Thanks,
      Jess

**Comment by ge0rg** [comment]
It would be nice if Kenwood would actually allow proper remote control of the TNC in APRS mode, so you could inject packet payloads like a message.

That said, I just added a small workaround into the "Kenwood (NMEA waypoint)" decoder so that it understands TNC2 packets (plain ASCII) and strips any leading "cmd:" from them.

It will still emit NMEA packets, unless you disable GPS on your Android device or disconnect the TX line on the serial side of things.

Could you please check if this APK works for you? http://aprsdroid.org/download/builds/APRSdroid-gl-2017-03-02.apk

**Comment by jesshaas** [comment]
I will try it out first thing tomorrow. 

Dont get me started on things that would be nice if Kenwood would allow with this radio. Even switching bands via the serial port requires a hack.

**Comment by jesshaas** [comment]
It works perfectly even with aprsdroid spamming it with all the GPS data(which the TNC command prompt is probably responding to every time). I have only tested it for a few minutes but every packet that the D710 decoded showed up in aprsdroid. I was even seeing myself when my packets bounced off a digipeater so it doesn't seem to have problems when the radio transmits packets. I am a happy man. Now just need to get packets transmitting so we can send messages ;-)

**Comment by jesshaas** [comment]
After some more use I did notice a few packets that the D710 recognized that aprsdroid didnt. Not sure if things got confused from the GPS spamming, they were packets aprsdroid didny understand for some reason or if it was some other issue.

**Comment by LA3QMA** [comment]
Use the radio in APRS mode with the APRS output menu on,
When aprsdroid or other software needs to transmit data switch to the TNC mode using the TC and PS commands.

This doesn't take 3 seconds. But when the software puts the radio in TNC mode you don't receive anything on the display on the radio until the software sends the correct TC/PS commans again.

Using this method you get the APRS displayed on the radio and also the software supporting this hack/hybrid mode.

https://github.com/LA3QMA/TM-V71_TM-D710-Kenwood

73 de LA3QMA

**Comment by jesshaas** [comment]
APK you provided is working great but I need a version with offline maps. Any chance of this support being added to the github repo anytime soon?

Thanks,
   Jess

**Comment by ge0rg** [comment]
Could you please try https://aprsdroid.org/download/builds/APRSdroid-gl-2017-04-26_osm.apk - this is an experimental build created with an updated Android SDK, and it might have funny side effects.

**Comment by ge0rg** [comment]
Some more fixes and a GPS on/off switch in APRSdroid 1.4rc3, available from https://play.google.com/apps/testing/org.aprsdroid.app and as a nightly APK from the usual location.

The newer Kenwoods have a pletohra of APRS interop modes, and they all suck ;) If one day, a Kenwood engineer responds to my inquiries, I'll have a look at implementing a shared-operation APRS setup with a sane protocol, but for now I'll leave the focus at either of these:

* Kenwood in KISS TNC mode, all APRS happens on the Android
* Kenwood in APRS mode, APRSdroid shows the map and optionally provides GPS

I still have no good feeling for a mode where APRSdroid switches the D7x0 between KISS and APRS mode for individual packets. Maybe this will become a thing together with rig control, but I'd consider it a WONTFIX for now.

**Comment by jesshaas** [comment]
That OSM build has been working great for me. Thanks!

What about a way for aprsdroid to recognize when the radio has been switched from aprs to kiss mode? Its just one button on the radio to switch but currently requires digging through the menus in aprsdroid to switch on its end. I just want it to be quick and simple to switch to a mode where i can use aprsdroid for sending messages.

**Comment by turbodb** [comment]
@jesshaas - how do you have the D710G communicating with APRSDroid - is it wired or bluetooth? I'd like to get the DATA port on the radio (not on the control panel) hooked to some Bluetooth "transmitter" that I then pair with a tablet running APRSDroid.

Right now I'm running mobilinkd for APRS, but like you, I'd like to use the built-in APRS on the Kenwood.

Ideas?

**Comment by jesshaas** [comment]
I have a bluetooth serial module on the TNC port. Also have another one on thebother serial port fr rig control and programming. Here is how i did it: http://jesshaas.com/projects/d710bt/ The bluetooth modules are cheap on ebay if your patient enough to wait for them to come from china. They just need a bit of hacking.

This setup with the Aprsdroid with offline maps works great.

**Comment by turbodb** [comment]
Thanks. When you say the TNC port - do you mean the one on the receiver labeled "DATA" or do you mean the one on the display (labeled "COM" I think). Do they output the same info? I want to use the "DATA" port, as my display is not in a location where I want to have wires hanging off of it.

Thanks again.

**Comment by ge0rg** [comment]
@turbodb the "DATA" port is an audio port without attenuation, where you can connect a TNC. The whole APRS functionality in the D7x0s is performed on the head unit, therefore you only connect a serial port to the head. Your options are:

- use the head unit's "COM" serial port to use the head unit as a TNC in `KISS` mode, APRSdroid is the only APRS station
- use the head unit's GPS port (2.5mm) in `APRS` mode, APRSdroid will only do map display and GPS
- use the "DATA" port on the back of the D7x0 to connect an audio cable to your Android or to the TNC. The head unit will be completely separate from that and can be used in whatever mode you wish. You can have APRS on the head unit and on APRSdroid with a Mobilinkd this way, just ensure that you have different SSIDs.

**Comment by turbodb** [comment]
@ge0rg - Ahh, got it - thanks.

I like bullet 3 - can I have APRS on the head unit and APRSdroid (with Mobillinkd) running on the same band, as long as I'm using different SSID's? (if I don't want to tie up both bands with APRS)

(I can just try this obviously, but curious if that's what you are saying in bullet #3)

**Comment by ge0rg** [comment]
I haven't tested this myself, and I don't know if the DATA port will always output one fixed bad or switch to the active one. In theory, it should work.

**Comment by LA3QMA** [comment]
If you are using the internal TNC the radio is transmitting on whatever
band the radio is configured to.
If you are using the DATA port it's using the band that you have selected
"ctrl" "ptt"

**Comment by jesshaas** [comment]
You can also use the COM port on the head unit with the radio in APRS mode. This gives you full receive functionality in aprsdroid.

**Comment by tylojuky** [comment]
@jesshass
When using the Com port on the head unit with the radio in APRS mode, will I get my own position on the APRSDroid map or only those I receive?

Also, has anyone tried connecting directly to Android via cable/OTG?

**Comment by jesshaas** [comment]
Aprsdroid won't see transmitted packets. If you set your callsign+ssid to the same as the radio is set to then whenever you are digipeated it will get your position. Distances to other stations, etc are very off until this happens.

**Comment by tylojuky** [comment]
@jesshass - thank you, thank you, thank you. 
It took me several tries before I finally figured out the difference between "GPS Port", "COM Port", and "PC Port" and where all the menus are controlling those on the D710G. I was about to give up and go to a Windows solution until I found you and your posts.
While waiting on the parts to build a serial-to-bluetooth adapter like you posted, I grabbed my RT Systems cable (the only one I have that's 8-pin mini DIN to USB) for one last trial. I connected that cable to the COM port on the back of the D710G head (the part with the display / not the radio where the coax connects) and to the OTG adapter that came with my Samsung Galaxy S8.
In Airplane mode (no internet connection) and with APRSDroid set to Kenwood/USB Serial/9600 baud I was able to see beacons in my area on the map (using the APK that uses map files stored on the phone) - this is a huge success and is exactly what I needed.
I never did see myself on the map(?) I wonder if the Kenwood is smart enough to filter received packets that it sent itself(?) - or at least not pass them through to the COM port.
In any case, I'm very happy to be able to do this on an Android device and no network connection. (I'm shopping for tablets now that I know this can work). This will be a huge benefit when we're working bike races in the hills of Oklahoma.
I'm still going to build a serial bluetooth adapter like you described and leave it tucked away under the dash because wireless is just way nicer.
I also happily paid $5 for the version of APRSDroid on the Play store - and I'll happily do it again when I get a tablet (even if I don't have to).

BTW - I know this is an "feature request" thread but I couldn't find anywhere else to post comments and questions like this. If there is a better place, please let me know. I'd love to read and contribute to the discussions.

Thanks again.
Kyle

**Comment by Cody0508** [comment]
@tylojuky I just received my RT systems cable in the mail today and I was trying to set up my 710g like you did on an old android phone. I was wanting to test out the mapping where it will show the position of aprs packets that are received on my radio. I haven’t had any luck getting them to show up. How do you have aprsdroid configured?  I think I have the radio set up right. I was wanting to try it hard wired without the Bluetooth option for now just to get it running.

**Comment by tylojuky** [comment]
That’s how I set it up at first as well. I don’t know what USB configuration options are available on your old android phone but you will most definitely need “full USB” capability. Not all phones have this. On my Samsung it is branded “USB-OTG” (for “on-the-go”) but it just actually means the USB port that is used to charge the phone and sync music and such is a fully-featured USB port that is accessible to applications for two-way serial communications - just like on your computer. If your phone will support this then you can get it working. If not, you’ll need a different Android device.

The biggest confusion for me was all the “ports” that the 710g has available and how they’re named/where their settings are in the menus and where they physically are on the radio. The only one you need to worry about is the “COM” port - not "PC port", not "Ext Data", not “TNC", not “GPS Port”, not anything else - just “COM”. The physical port is a mini-DIN port on the back side of the 710g control head. Leave all other port/baud speed settings alone. The COM port speed is found in the AUX submenu - number 929. This will need to be set to match the USB Serial baud rate in APRSDROID. Everything else for getting up and going on the 710g is in the APRS submenu. 

I’ve attached pictures of the relevant menu settings to this email. Substitute your call sign and a few other personal choice settings.
Once those are setup, set side-A to 144.390 simplex, make sure GPS is on and acquired (“GPS” is flashing) and hit the TNC button one time (“APRS12” will display).

I also attached a picture of my APRSDROID settings. Walk through each of these too.

Using your RT cable, connect the COM port on the back of the 710g head unit to your android USB port (need a USB adapter).

Hit “Start Tracking” in APRSDROID and you’re off and running.
You should see the same entries in APRSDROID Hub as you see in the received message list on the 710g
If you want to be sure what you see in APRSDROID is coming from the radio, turn off cellular data and wifi on your phone and make sure you’re still getting beacon updates from the 710g.

One thing I noticed right away is the “Filter” settings on the 710g don’t appear to change what gets sent over the COM port. APRSDROID gets every beacon received by the 710g - even if you have it set to filter out some of them.

Let me know if you have specific questions

Kyle


On May 22, 2018, at 10:10 PM, Cody0508 <notifications@github.com> wrote:

@tylojuky <https://github.com/tylojuky> I just received my RT systems cable in the mail today and I was trying to set up my 710g like you did on an old android phone. I was wanting to test out the mapping where it will show the position of aprs packets that are received on my radio. I haven’t had any luck getting them to show up. How do you have aprsdroid configured? I think I have the radio set up right. I was wanting to try it hard wired without the Bluetooth option for now just to get it running.

**Comment by Cody0508** [comment]
@tylojuky Thanks for all of that info Kyle.  

The phone I am using allows debugging and data transfer in the "developer" options so I think that should be fine.  

I feel sure the radio is set up properly and I am using the correct "COM" port on the back of the display as you mentioned.  I am just wondering if the issue I'm having is with APRSDROID.  

I didn't see any pictures attached to the post/email that you mentioned. If you could just try and send the setup pictures you have of APRSDROID, that would be great, and I believe get me fixed up.  I feel like I have been over everything pretty thoroughly but I must be missing something.  

Thanks for your time!

Cody

**Comment by tylojuky** [comment]
I've attached a Pic of my setup screen for APRSDroid. One thing: you must use "Kenwood" protocol on both ALRSDroid and the radio. Other than that it's a pretty straightforward setup.   
  
For me, it sometimes takes a mini or two after I start the TNC before I start seeing beacons on the map so be patient.   
  
  
  
  
 Kyle Hager Sent from a mobile device. Please excuse typos.  

  
  
  
  
  
>

**Comment by matobler** [comment]
@ge0rg thanks for all those updates. I am currently using APRSDroid connected to the COM port (bluetooth) with the Kenwood in APRS mode and its working great. I was trying to figure out how to connect through the COM port with Kenwood in KISS TNC mode so that all the APRS happens on the tablet and I can send messages. So far no success. How do I have APRSDroid submit the commands to put the TNC in KISS mode? Or should the TNC go into KISS mode automatically when I press the TNC button? Tried that but did not work.

**Comment by matobler** [comment]
Ok, I finally figured it out. For others who might be looking for the same information, under TNC init string put (on two lines):

KISS ON
RESTART

after that everything works fine. 

@jesshaas, thanks for the write-up on the bluetooth module. I got the same module but removed the serial connector and was able to solder it directly to the COM port and put it inside the head unit. Also put a small GPS receiver in there which turned my D710A into a D710GA with integrated bluetooth and no external wires. Pretty nice solution and works great with APRSDroid.

**Comment by matobler** [comment]
Here it goes.  Did not realize I could attach pictures here. GPS module on the left, bluetooh in the center. Tight fit but it works. They are soldered to the PCB where the connectors are soldered on. There is a place to get 5V. Details on how to connect the GPS can be found in the link below. I got a GPS module with RS232 not TTL. Bluetooth is the same idea, just soldering to the COM port. RTS and DTS pins need to be bridged to be able to send data as @jesshaas has pointed out in his write-up (I used a small wire across the pins in the COM port connector so I can remove the bridge if I want to use a serial cable at some point). Bluetooth turns on when the radio turns on and waits for a connection. One can even use the serial cable to re-program the bluetooth module if needed.

http://www.w2dhs.com/2016/07/07/adding-gps-to-a-kenwood-rc-d710-and-possibly-tm-d710/

![image](https://user-images.githubusercontent.com/2934789/44161663-daa89f80-a072-11e8-9d1d-48953e3f56cf.png)

![image](https://user-images.githubusercontent.com/2934789/44162017-d92ba700-a073-11e8-988e-35a35656e1c6.png)

**Comment by matobler** [comment]
@ge0rg feel free to use the pictures in the Wiki. That will be a really great resource. From what I can tell the COM port outputs the full APRS packets (more information than the GPS port). In the 3rd post @jesshaas indicates that you can switch TNC mode through the COM port but I personally have not tried that. I can try it out.

**Comment by matobler** [comment]
I did some tests last night. You can switch between **APRS** and **PACKET** mode through the **COM** port and from **PACKET** to **KISS**. Attached a log showing that. I started out in **APRS** mode, switched to **PACKET** and then **KISS** and then back to **APRS**. This was all done over bluetooth through the **COM** port on the head with a serial terminal. Commands are as described by @jesshaas  earlier. I run APRS on Band B so it's TN1,1 instead of TN 1,0 (Band A).

[serial_20180815_230643.txt](https://github.com/ge0rg/aprsdroid/files/2295114/serial_20180815_230643.txt)

**Comment by tylojuky** [comment]
@matobler - Very nice!!! I have have the "G" model already but hiding the bluetooth transmitter inside the head is genius. Very nice.

Are you able to send APRS messages using APRSDroid with the Kenwood?
If so, are you using "Kenwood (NMEA waypoint)" protocol or something else?

I was looking around for a way to configure APRSDroid to switch modes when sending a message then switch back again --- but maybe I'm making this too complicated. If I'm using "Kenwood" protocol, does messaging just work?

**Comment by tylojuky** [comment]
@matobler - I deleted my first question because I re-read your posting about using KISS mode and "...all the APRS happens on the tablet..."

So I have a new question:
When you're in KISS mode (and you send the correct TNC init string to the Kenwood), I assume messaging works (please verify).

What, exactly do you mean by "all the APRS happens on the tablet" - Specifically, does the APRS beacon go out over your cellular data connection does the Kenwood send it? If the radio, is this done "on command" from APRSDroid or using Kenwood's normal APRS logic???? Lots of questions about this so maybe just start simple(?)

**Comment by matobler** [comment]
When you are in KISS mode the radio does not parse any APRS packages. That means you don't see stations or messages on the radio head unit. Everything is done by APRSDroid. APRSDroid controls the radio so beacons go out over the radio, no cellular data connection needed. And yes, messaging works fine through APRSDroid in KISS mode.

**Comment by tylojuky** [comment]
@matobler  - Thanks for that. It makes sense to me now. I'll try it out next time I'm running APRS (or maybe sooner for fun).

**Comment by squaremining** [comment]
@ge0rg Wondering if you can point me in the right direction. I wired up a HC06 to my head unit. Pulled the 5v power and used the TX/RX pins off the head unit com port. The radio is a newer Kenwood D710. I've set up APRSdroid as per your instructions on the WIKI re: Kenwood D7x0.

Have tried to jump the pins noted above. Bottom two on the com port. Have tried without. Bluetooth monitor shows nothing for a long time and then gives a page or so of just random garbage. Baud rate is set to 9600. Com port is open on the APRS menu of the radio. 

APRSdroid shows "connected to TNC" in the log. However it does not get any packets when I do testing. I have the same callsign set on APRSdroid, but using a -10 on the app and a -9 on the radio. I have another handheld set to -7 and nothing shows in the logs whatsoever other than the occasional -10 log from the android app with the status "Bluetooth OK".

Am I missing something in this setup?

**Comment by ge0rg** [comment]
@squaremining did you also connect gnd? It looks like the serial data isn't arriving properly, you should check the cabling and the HC06 adapter. KISS will look like random garbage with the plaintext parts interspersed, but in monitor mode you should see proper ASCII output. Either way, it should come immediately and not in busts.

**Comment by matobler** [comment]
Did you add a TTL to RS232 converter. The interface on the radio is RS232 while the HC06 uses TTL they are not compatible.

**Comment by BI1IZF** [comment]
![QQ截图20220507231540](https://user-images.githubusercontent.com/104998180/167260503-46d26dfc-245f-4997-bdd3-5fdc040d7e06.jpg)
I'm designing a module that has bluetooth and GPS and I'm still testing it.
